src/optix_coop_vec.cpp

```diff
@@ -0,0 +1,488 @@
+                copy_from_buffer("dst", v);
+            }
+            break;
+
+        case VarKind::CoopVecBinaryOp:
+            fmt("        .reg.b32 %op, %type, %size, %u;\n"
+                "        mov.b32 %op, $u;\n"
+                "        mov.b32 %type, $u;\n"
+                "        mov.b32 %size, $u;\n",
+                jitc_optix_coop_vec_op_id((JitOp) v->literal),
+                jitc_optix_coop_vec_type_id((VarType) v->type),
+                length
+            );
+
+            if (reg_count) {
+                put("        call (");
+                put_elem(v, reg_count, false, true);
+                fmt("), _optix_vector_op2_$uxi32, (%op, %type, %size, %type, %size, ", reg_count);
+                put_elem(a0, reg_count, true);
+                put_elem(a1, reg_count, false);
+                put(");\n");
+            } else {
+                declare_buffer("src_0", a0);
+                declare_buffer("src_1", a1);
+                declare_buffer("dst", v);
+                copy_to_buffer("src_0", a0);
+                copy_to_buffer("src_1", a1);
+                put("        call (), _optix_vector_op2_ptr, (%op, %type, %size, %type, %size, %src_0, %src_1, %dst);\n");
+                copy_from_buffer("dst", v);
+            }
+            break;
+
+        case VarKind::CoopVecTernaryOp:
+            fmt("        .reg.b32 %op, %type, %size, %u;\n"
+                "        mov.b32 %op, $u;\n"
+                "        mov.b32 %type, $u;\n"
+                "        mov.b32 %size, $u;\n",
+                jitc_optix_coop_vec_op_id((JitOp) v->literal),
+                jitc_optix_coop_vec_type_id((VarType) v->type),
+                length
+            );
+
+            if (reg_count) {
+                put("        call (");
+                put_elem(v, reg_count, false, true);
+                fmt("), _optix_vector_op3_$uxi32, (%op, %type, %size, %type, %size, ", reg_count);
+                put_elem(a0, reg_count, true);
+                put_elem(a1, reg_count, true);
+                put_elem(a2, reg_count, false);
+                put(");\n");
+            } else {
+                declare_buffer("src_0", a0);
+                declare_buffer("src_1", a1);
+                declare_buffer("src_2", a2);
+                declare_buffer("dst", v);
+                copy_to_buffer("src_0", a0);
+                copy_to_buffer("src_1", a1);
+                copy_to_buffer("src_2", a2);
+                put("        call (), _optix_vector_op3_ptr, (%op, %type, %size, %type, %size, %src_0, %src_1, %src_2, %dst);\n");
+                copy_from_buffer("dst", v);
+            }
+            break;
+
+        case VarKind::Bitcast:
+                for (uint32_t i =  0; i < reg_count; ++i)
+                    fmt("        mov.b32 %cv$u_$u, %cv$u_$u;\n",
+                        v->reg_index, i, a0->reg_index, i);
+            break;
```

Comment:
Isn't this already handled in `cuda_eval.cpp`? Is it a different op than the existing `jitc_var_cast()`?

Reply:
It's a bit-cast for coop vectors, which are distributed over multiple registers. The code that needs to be generated in this case is different from what is in `cuda_eval.cpp` (all OptiX coopvec-related code generation is in `optix_coop_vec.cpp` and not in `cuda_eval.cpp`).

JIT variables use the `jit_var_*` API. Cooperative vectors use the `jit_coop_vec_*` API. I added a clarification to `jit.h`.